subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java

```diff
@@ -93,16 +102,67 @@ public void apply(Project project) {
      * Creates the configurations used by plugin.
      */
     private void addJacocoConfigurations() {
-        Configuration agentConf = project.getConfigurations().create(AGENT_CONFIGURATION_NAME);
+        ConfigurationContainer configurations = project.getConfigurations();
+
+        Configuration agentConf = configurations.create(AGENT_CONFIGURATION_NAME);
         agentConf.setVisible(false);
         agentConf.setTransitive(true);
         agentConf.setDescription("The Jacoco agent to use to get coverage data.");
         deprecateForConsumption(agentConf);
-        Configuration antConf = project.getConfigurations().create(ANT_CONFIGURATION_NAME);
+        Configuration antConf = configurations.create(ANT_CONFIGURATION_NAME);
         antConf.setVisible(false);
         antConf.setTransitive(true);
         antConf.setDescription("The Jacoco ant tasks to use to get execute Gradle tasks.");
         deprecateForConsumption(antConf);
+
+        Configuration aggregatedJacocoReport = configurations.create(JacocoAggregatedReport.AGGREGATION_CONFIGURATION_NAME, c -> {
+            c.setVisible(false);
+            c.setCanBeResolved(false);
+            c.setCanBeConsumed(false);
+            c.setDescription("Bucket for project dependencies to aggregate test coverage data from.");
+        });
+
+        project.getPluginManager().withPlugin("java", plugin -> {
+            JavaPluginExtension javaPluginExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+            // TODO: what about other "production" source sets that the users might add?
+            SourceSet mainSourceSet = javaPluginExtension.getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+            Configuration implementation = configurations.getByName(mainSourceSet.getImplementationConfigurationName());
+            aggregatedJacocoReport.extendsFrom(implementation);
+
+            // TODO: this probably belongs in java plugin
+            Configuration sourceDirectoriesElements = createVariant("sourceDirectoriesElements",
+                JacocoAggregatedReport.sourceDirectoriesAttributes(project));
+            sourceDirectoriesElements.extendsFrom(implementation);
+            sourceDirectoriesElements.setDescription("Java source directories variant.");
+            mainSourceSet.getJava().getSrcDirs().forEach(f -> sourceDirectoriesElements.getOutgoing().artifact(f));
+
+            TaskContainer tasks = project.getTasks();
+            project.afterEvaluate(p -> {
```

Comment:
I hope there is, but so far this is the best I was able to come up with without having all the Test tasks eagerly realized.

The problem is that before the project is evaluated, we do not know the names of all the test tasks. `tasks.withType(Test.class)` yields a `TaskCollection`, which is not a provider and which is empty before the project is evaluated.
I register a variant for each test task and use the task name as an attribute which can then be used to aggregate different types of tests. Even if this attribute was something else than task name, I still need to do this for every test task without realizing that task.

Reply:
Then I believe that the recommended pattern is `tasks.withType(Test.class).configureEach { ... }` which will be lazy. That is it will be invoked only if the a test task us realized and added to the collection.